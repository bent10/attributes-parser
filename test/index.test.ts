/// <reference types="vitest/globals" />

import {
  tokenizeAttrs,
  parseAttrs,
  serializeTokens,
  type Token
} from '../src/index.js'

describe('tokenizeAttrs', () => {
  it('should tokenize a string of attributes', () => {
    const input = `id="foo" class=\'bar\' num="3.14" data-value=baz name="@myName" data-value="[1, 2, 3]" fooBar="{foo: 'bar'}" checked=false disabled`

    const result = tokenizeAttrs(input)
    expect(result).toMatchSnapshot()
  })
})

describe('parseAttrs', () => {
  it('should parse attributes into key-value pairs', () => {
    const input = `id="foo" class=\'bar\' num="-3.14" data-value=baz name="@myName" data-value="[1, 2, 3]" fooBar="{foo: 'bar'}" checked=false disabled`
    const expectedParsedAttrs = {
      id: 'foo',
      class: 'bar',
      num: -3.14,
      'data-value': [1, 2, 3],
      name: '@myName',
      fooBar: { foo: 'bar' },
      checked: false,
      disabled: true
    }

    const result = parseAttrs(input)
    expect(result).toEqual(expectedParsedAttrs)
  })

  it('should use the last duplicate key value', () => {
    const input = 'foo="foo" foo="bar"'

    const result = parseAttrs(input)
    expect(result).toEqual({ foo: 'bar' })
  })

  it('should handle number-like correctly', () => {
    const input = 'foo="42" bar="-42" baz="3.14" qux="0.5" quux="-0.5" nan=".5"'
    const expectedParsedAttrs = {
      foo: 42,
      bar: -42,
      baz: 3.14,
      qux: 0.5,
      quux: -0.5,
      nan: '.5'
    }

    const result = parseAttrs(input)
    expect(result).toEqual(expectedParsedAttrs)
  })

  it('should handle double-quoted values with special characters', () => {
    const input =
      'name="@myName" data-value="[1, 2, 3]" foo="{bar: \'bar\'}" checked=false'
    const expectedParsedAttrs = {
      name: '@myName',
      'data-value': [1, 2, 3],
      foo: { bar: 'bar' },
      checked: false
    }

    const result = parseAttrs(input)
    expect(result).toEqual(expectedParsedAttrs)
  })

  it('should handle unquoted values with special characters', () => {
    const input = 'name=myName data-value=[1,2,3] disabled={value:true}'
    const expectedParsedAttrs = {
      name: 'myName',
      'data-value': [1, 2, 3],
      disabled: { value: true }
    }

    const result = parseAttrs(input)
    expect(result).toEqual(expectedParsedAttrs)
  })
})

describe('serializeTokens', () => {
  it('should serialize an array of tokens into a string', () => {
    const tokens = [
      { type: 'name', value: 'id', text: 'id' },
      { type: 'separator', value: '=', text: '=' },
      { type: 'doubleQuotedvalue', value: 'foo', text: '"foo"' }
    ] as Token[]
    const expectedSerializedString = 'id="foo"'

    const result = serializeTokens(tokens)
    expect(result).toEqual(expectedSerializedString)
  })

  it('should handle tokens with different types', () => {
    const tokens = [
      { type: 'name', value: 'id', text: 'id' },
      { type: 'separator', value: '=', text: '=' },
      { type: 'doubleQuotedvalue', value: 'foo', text: '"foo"' },
      { type: 'invalidType', value: 'ignored', text: 'ignored' }, // ignored silently
      { type: 'whitespace', value: ' ', text: ' ' },
      { type: 'name', value: 'class', text: 'class' },
      { type: 'separator', value: '=', text: '=' },
      {
        type: 'singleQuotedvalue',
        value: { bar: 'bar' },
        text: `'{bar: "bar"}'`
      }
    ] as Token[]
    const expectedSerializedString = `id="foo" class='{bar: "bar"}'`

    const result = serializeTokens(tokens)
    expect(result).toEqual(expectedSerializedString)
  })
})

it('test', () => {
  const attrs = `id="foo" class=\'bar\' num="123" data-value=baz name="@myName" data-value="[1, 2, 3]" fooBar="{foo: 'bar'}" checked=false disabled`
  const parsedAttrs = parseAttrs(attrs)

  console.log(parsedAttrs)
})
