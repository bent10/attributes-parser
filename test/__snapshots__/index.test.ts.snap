// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tokenizeAttrs > should tokenize a string of attributes 1`] = `
[
  {
    "col": 1,
    "line": 1,
    "lineBreaks": 0,
    "offset": 0,
    "text": "id",
    "toString": [Function],
    "type": "name",
    "value": "id",
  },
  {
    "col": 3,
    "line": 1,
    "lineBreaks": 0,
    "offset": 2,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 4,
    "line": 1,
    "lineBreaks": 0,
    "offset": 3,
    "text": "\\"foo\\"",
    "toString": [Function],
    "type": "doubleQuotedvalue",
    "value": "foo",
  },
  {
    "col": 9,
    "line": 1,
    "lineBreaks": 0,
    "offset": 8,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 10,
    "line": 1,
    "lineBreaks": 0,
    "offset": 9,
    "text": "class",
    "toString": [Function],
    "type": "name",
    "value": "class",
  },
  {
    "col": 15,
    "line": 1,
    "lineBreaks": 0,
    "offset": 14,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 16,
    "line": 1,
    "lineBreaks": 0,
    "offset": 15,
    "text": "'bar'",
    "toString": [Function],
    "type": "singleQuotedvalue",
    "value": "bar",
  },
  {
    "col": 21,
    "line": 1,
    "lineBreaks": 0,
    "offset": 20,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 22,
    "line": 1,
    "lineBreaks": 0,
    "offset": 21,
    "text": "num",
    "toString": [Function],
    "type": "name",
    "value": "num",
  },
  {
    "col": 25,
    "line": 1,
    "lineBreaks": 0,
    "offset": 24,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 26,
    "line": 1,
    "lineBreaks": 0,
    "offset": 25,
    "text": "\\"3.14\\"",
    "toString": [Function],
    "type": "doubleQuotedvalue",
    "value": 3.14,
  },
  {
    "col": 32,
    "line": 1,
    "lineBreaks": 0,
    "offset": 31,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 33,
    "line": 1,
    "lineBreaks": 0,
    "offset": 32,
    "text": "data-value",
    "toString": [Function],
    "type": "name",
    "value": "data-value",
  },
  {
    "col": 43,
    "line": 1,
    "lineBreaks": 0,
    "offset": 42,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 44,
    "line": 1,
    "lineBreaks": 0,
    "offset": 43,
    "text": "baz",
    "toString": [Function],
    "type": "unquotedvalue",
    "value": "baz",
  },
  {
    "col": 47,
    "line": 1,
    "lineBreaks": 0,
    "offset": 46,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 48,
    "line": 1,
    "lineBreaks": 0,
    "offset": 47,
    "text": "name",
    "toString": [Function],
    "type": "name",
    "value": "name",
  },
  {
    "col": 52,
    "line": 1,
    "lineBreaks": 0,
    "offset": 51,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 53,
    "line": 1,
    "lineBreaks": 0,
    "offset": 52,
    "text": "\\"@myName\\"",
    "toString": [Function],
    "type": "doubleQuotedvalue",
    "value": "@myName",
  },
  {
    "col": 62,
    "line": 1,
    "lineBreaks": 0,
    "offset": 61,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 63,
    "line": 1,
    "lineBreaks": 0,
    "offset": 62,
    "text": "data-value",
    "toString": [Function],
    "type": "name",
    "value": "data-value",
  },
  {
    "col": 73,
    "line": 1,
    "lineBreaks": 0,
    "offset": 72,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 74,
    "line": 1,
    "lineBreaks": 0,
    "offset": 73,
    "text": "\\"[1, 2, 3]\\"",
    "toString": [Function],
    "type": "doubleQuotedvalue",
    "value": [
      1,
      2,
      3,
    ],
  },
  {
    "col": 85,
    "line": 1,
    "lineBreaks": 0,
    "offset": 84,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 86,
    "line": 1,
    "lineBreaks": 0,
    "offset": 85,
    "text": "fooBar",
    "toString": [Function],
    "type": "name",
    "value": "fooBar",
  },
  {
    "col": 92,
    "line": 1,
    "lineBreaks": 0,
    "offset": 91,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 93,
    "line": 1,
    "lineBreaks": 0,
    "offset": 92,
    "text": "\\"{foo: 'bar'}\\"",
    "toString": [Function],
    "type": "doubleQuotedvalue",
    "value": {
      "foo": "bar",
    },
  },
  {
    "col": 107,
    "line": 1,
    "lineBreaks": 0,
    "offset": 106,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 108,
    "line": 1,
    "lineBreaks": 0,
    "offset": 107,
    "text": "checked",
    "toString": [Function],
    "type": "name",
    "value": "checked",
  },
  {
    "col": 115,
    "line": 1,
    "lineBreaks": 0,
    "offset": 114,
    "text": "=",
    "toString": [Function],
    "type": "separator",
    "value": "=",
  },
  {
    "col": 116,
    "line": 1,
    "lineBreaks": 0,
    "offset": 115,
    "text": "false",
    "toString": [Function],
    "type": "unquotedvalue",
    "value": false,
  },
  {
    "col": 121,
    "line": 1,
    "lineBreaks": 0,
    "offset": 120,
    "text": " ",
    "toString": [Function],
    "type": "whitespace",
    "value": " ",
  },
  {
    "col": 122,
    "line": 1,
    "lineBreaks": 0,
    "offset": 121,
    "text": "disabled",
    "toString": [Function],
    "type": "name",
    "value": "disabled",
  },
]
`;
